### ISSUE TITLE
[red-knot] support invalid syntax without panics

### ISSUE THREAD

**Issue by carljm** [astral-sh/ruff#13778]
The ruff parser is error-resilient and will generate a best-effort AST for any input, but red-knot currently assumes a valid AST in some places, and this can cause panics if it's run over an invalid AST.

We should do best-effort type-checking for invalid ASTs, and never panic.

The trick is to do this while, as much as feasible, still maintaining some useful internal invariants, and failing fast (panic is good in this case!) if those invariants are violated, rather than allowing programming errors to pass silently and result in type-checking bugs.

**Comment by carljm** [comment]
https://github.com/astral-sh/ruff/pull/13701 is a useful reference on what changes were needed as of a few days ago to avoid panics on the Python files present in the ruff repo. At least some of those are due to invalid syntax.

**Comment by carljm** [comment]
In addition to "not panic", we also want to avoid useless or confusing diagnostics.

Inferring useful types in the face of syntax errors is less of a priority.

**Comment by sharkdp** [comment]
We can now run red knot on all invalid-syntax examples in ruffs repository without panicking. We can also run it on a fuzzing corpus with ~4000 invalid-syntax files (https://github.com/astral-sh/ruff/issues/13448). The next step could be to look into direct fuzzing (#14157), or to look for other corpuses (corpi? corpora?) with invalid syntax examples.

**Comment by MichaReiser** [comment]
One specific case that we should handle is that we don't put names synthesized by the parser during recovery into the symbol table (or try to look them up). There's an AST method that allows testing for whether the name is valid.

**Comment by dhruvmanila** [comment]
I've been exploring ways to create panics on source code with syntax errors but I haven't found any that requires any source code that's raising syntax errors which isn't already documented which is https://github.com/astral-sh/ruff/issues/14672 and https://github.com/astral-sh/ruff/blob/b63c2e126b928fab4180eda5d9132552beeee7fe/crates/red_knot_workspace/tests/check.rs#L273-L277.

I've also ran the fuzzer which produces source code with invalid syntax (https://github.com/astral-sh/ruff/pull/14678) multiple times and it always finds the one as linked above. The corpus is initiated with most of the Python files in the Ruff code base and all Python files in the CPython code base. The fuzzer runs on them and then starts shuffling the bytes around producing new source code which is then filtered to avoid the ones that doesn't produce syntax errors as per our parser. This has an obvious limitation that the syntax errors that's raised by the compiler

I looked at https://github.com/astral-sh/ruff/issues/13778#issuecomment-2479383240 as well but I've been unable to create a source code that could panic. I think that might be because of lack of type inference for positions where that could occur like type parameters, function parameters in combination with the AST key that includes the node range. So, technically even though that behavior might be incorrect, it doesn't create a panic in red knot, at least not today.

Additionally, I also extracted all the examples from the syntax error test file in CPython code base (https://github.com/python/cpython/blob/bf21e2160d1dc6869fb230b90a23ab030835395b/Lib/test/test_syntax.py) as it's a doctest and ran red knot on top of it and got no panics. 

There are certain changes that can be done but I've been unable to create a source code related to those changes that would make red knot panic. I'm going to list them down here for posterity:
* Currently, for an invalid `Identifier`, the `ExprAttribute` doesn't set an `Invalid` expression context. We use this information to avoid recording a use of a symbol and / or creating the definition. Carl did mention previously that an attribute expression will be handled in a different manner and wouldn't create a definition.
* The parser recovers from a syntax error where a keyword is used instead of a name by creating a name expression with the keyword: https://play.ruff.rs/cb616a58-855a-431e-8ee3-3ecd6320e45d. The downstream tools have no way of knowing whether this identifier was recovered from this syntax error. Regardless, this shouldn't cause any panic.
* Avoid adding the symbol for an `ExprName` which has an invalid expression context as that indicates a syntax error from which the parser recovered

That said, I think we should periodically run the fuzzer and iteratively fix any errors that the model encounters. The model is still a work in progress so there are missing pieces which might prove necessary to create these panics if any.

**Comment by MichaReiser** [comment]
Thanks for writing this up. 

Do you think it would be possible and useful to have mdtests for invalid identifiers and keywords used as identifiers?

**Comment by dhruvmanila** [comment]
> Do you think it would be possible and useful to have mdtests for invalid identifiers and keywords used as identifiers?

I think they should be covered by the invalid parser tests:
* Keyword assignment target: https://github.com/astral-sh/ruff/blob/5137fcc9c81610f687b6cb45413ef83c2c5eea73/crates/ruff_python_parser/resources/inline/err/assign_stmt_keyword_target.py
* Invalid assignment target: https://github.com/astral-sh/ruff/blob/5137fcc9c81610f687b6cb45413ef83c2c5eea73/crates/ruff_python_parser/resources/invalid/statements/invalid_assignment_targets.py
* Invalid augmented assignment target: https://github.com/astral-sh/ruff/blob/5137fcc9c81610f687b6cb45413ef83c2c5eea73/crates/ruff_python_parser/resources/invalid/statements/invalid_augmented_assignment_target.py

Although, I think they don't have complete coverage especially around "keywords used as identifiers", I'll add them.

**Comment by MichaReiser** [comment]
That makes sense, thanks. 

We could also assert that red knot doesn't generate useless diagnostics for missing identifiers. For example. Red Knot should not emit a `undefined-variable` diagnostic for

```py
5 + 
```
